### ISSUE TITLE
More realistic VDOM example

### ISSUE THREAD

**Issue by justinfagnani** [rricard/proposal-refcollection#10]
The example in the README shows a record representing the VDOM for a button, but it's assigned to a variable once, rather than produced by a function, so it misses showing usage for components and the suitability for diffing.

A more realistic example would be probably something like this:

```js
const Button = (label) => {
  const refsc = new RefCollection();
  return {
    refs: refc,
    vdom: #{
        type: "div",
        props: #{ id: "root" },
        children: #[
            #{
                type: "button",
                props: #{
                    onClick: refc.ref(function () {
                        alert("Clicked!");
                    }),
                },
                children: #[
                    refc.ref(label), // The label is changeable between renders
                ],
            },
        ],
    },
  };
}
```

Apps will have multiple renders with the same components. It'll be state-change / event driven, but it's basically this:

```js
// first render
vdomLib.render(Button('Hello'), document.body)
// second render
vdomLib.render(Button('Welcome'), document.body)
```

The key behavior we need is that the record is identical between calls to `Button()` so that the VDOM library does not need to deeply diff the two records:

```js
Button('Hello').vdom === Button('Welcome').vdom; // this must be true
```

**Comment by rricard** [comment]
Thanks Justin for the example, I am not sure the proposal in its current state would actually work but I do understand what you're trying to do with this. I need to think about this use case a bit more but we can probably make it work.

**Comment by jridgewell** [comment]
> I am not sure the proposal in its current state would actually work

Can you explain? I think this works great as is.

The example here just makes it more clear that the records aren't being generated via some imperative code (eg, `a === b`, where `a` and `b` are coming from two separate source locations). Instead, it's usually a single function (representing a single source location) that generates a static record shape with new data inputs.

**Comment by justinfagnani** [comment]
I think the key question raised for me is what is the effect on equality of references?

For template literals, the values have no effect on the equality of the strings array. That would be ideal to me: references have no effect on equality, ie: every reference is equal to every other reference for the purposes of record identity.

I think the naive way I use RefCollection above won't actually work because the equality checks will fail.

So code like this will have to take great care to reuse reference Symbols, but because of the restrictions on passing Symbols to `ref()` (you can't have the same object multiple times in the RefCollection, the Symbol can't be pointing to an object already, etc.) I don't think you could really even use RefCollections.

The impedance mismatch is that for VDOM you really want the reference to identify the "hole" uniquely, not the value, and across evaluations A single value might be used in multiple locations in the record, so they should be allowed in multiple refs. A "hole" might be updated across renders to contain a different value, but that should not effect equality.

**Comment by rricard** [comment]
> I think the naive way I use RefCollection above won't actually work because the equality checks will fail.

Yes, I am under the impression and given previous conversations we had that you consider RefCollection as a "hole" system while we just define it as a weak symbol to object mapper. I don't think you need any extra proposal to hold holes. All you need is to keep a record/tuple struct and an associated object mapping symbols to object that does not need to be weak:

```js
const labelRef = Symbol();
const clickRef = Symbol();
const Button = (label) => {
  const refs = {
    [labelRef]: label,
    [clickRef]: function () {
      alert("Clicked!");
    },
  };
  return {
    refs,
    vdom: #{
        type: "div",
        props: #{ id: "root" },
        children: #[
            #{
                type: "button",
                props: #{
                    onClick: clickRef,
                },
                children: #[
                    labelRef,
                ],
            },
        ],
    },
  };
}

// first render
vdomLib.render(Button('Hello'), document.body)
// second render
vdomLib.render(Button('Welcome'), document.body)

Button('Hello').vdom === Button('Welcome').vdom; // this must be true

**Comment by justinfagnani** [comment]
Yeah, I was hoping this proposal would make this pattern more ergonomic. Something like the indexing system from the playground a while back.

In this case I think I would remove the VDOM example from the readme, and maybe even specifically say that VDOM is not one of the use cases solved by this proposal.

I'd be very interested in what use cases this proposal does target.

**Comment by rricard** [comment]
> In this case I think I would remove the VDOM example from the readme, and maybe even specifically say that VDOM is not one of the use cases solved by this proposal.

Yea, now that I understand better what's going on here I think that's the right call. I'll change that next time I get to have time for a rewrite. Thank you for making things clear on this. I'm interested in looking in the vdom use case but I have to understand it better first!
